lib/ipam/ipam.go

```diff
@@ -101,27 +101,139 @@ func (c ipamClient) AutoAssign(ctx context.Context, args AutoAssignArgs) ([]net.
+		_, err := c.client.Delete(ctx, aff.Key, aff.Revision)
+		if err != nil {
+			return nil, err
+		}
+		return nil, errors.New("Affinity is stale")
+	}
+
+	// If the block does match the affinity but the affinity has not been confirmed,
+	// try to confirm it.
+	if state != model.StateConfirmed {
+		// Write the affinity as pending.
+		aff.Value.(*model.BlockAffinity).State = model.StatePending
+		if err != nil {
+			return nil, err
+		}
+
+		// CAS the block to get a new revision and invalidate any other instances
+		// that might be trying to operate on the block.
+		b, err = c.client.Update(ctx, b)
+		if err != nil {
+			return nil, err
+		}
+
+		// Confirm the affinity.
+		aff.Value.(*model.BlockAffinity).State = model.StateConfirmed
+		aff, err = c.client.Update(ctx, aff)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return b, nil
+}
+
 func (c ipamClient) autoAssign(ctx context.Context, num int, handleID *string, attrs map[string]string, pools []net.IPNet, version ipVersion, host string) ([]net.IP, error) {
 
 	// Start by trying to assign from one of the host-affine blocks.  We
 	// always do strict checking at this stage, so it doesn't matter whether
 	// globally we have strict_affinity or not.
-	log.Debugf("Looking for addresses in current affine blocks for host '%s'", host)
+	logCtx := log.WithFields(log.Fields{"host": host})
+	logCtx.Debugf("Looking for addresses in current affine blocks for host")
 	affBlocks, err := c.blockReaderWriter.getAffineBlocks(ctx, host, version, pools)
 	if err != nil {
 		return nil, err
 	}
-	log.Debugf("Found %d affine IPv%d blocks for host '%s': %v", len(affBlocks), version.Number, host, affBlocks)
+	logCtx.Debugf("Found %d affine IPv%d blocks for host: %v", len(affBlocks), version.Number, affBlocks)
 	ips := []net.IP{}
+	newIPs := []net.IP{}
 	for len(ips) < num {
 		if len(affBlocks) == 0 {
-			log.Infof("Ran out of existing affine blocks for host '%s'", host)
+			logCtx.Infof("Ran out of existing affine blocks for host")
 			break
 		}
 		cidr := affBlocks[0]
 		affBlocks = affBlocks[1:]
-		ips, _ = c.assignFromExistingBlock(ctx, cidr, num, handleID, attrs, host, true)
-		log.Debugf("Block '%s' provided addresses: %v", cidr.String(), ips)
+
+		// Try to assign from this block - if we hit a CAS error, we'll try this block again.
+		// For any other error, we'll break out and try the next affine block.
+		for i := 0; i < ipamEtcdRetries; i++ {
+			// Get the affinity.
+			aff, err := c.client.Get(ctx, model.BlockAffinityKey{Host: host, CIDR: cidr}, "")
+			if err != nil {
+				logCtx.WithError(err).Warnf("Error getting affinity")
+				break
+			}
+
+			// Get the block which is referenced by the affinity, creating it if necessary.
+			b, err := c.getBlockFromAffinity(ctx, aff)
+			if err != nil {
+				// Couldn't get a block for this affinity.
+				logCtx.WithError(err).Warn("Couldn't get block for affinity")
+				if _, ok := err.(cerrors.ErrorResourceUpdateConflict); ok {
```

Comment:
Need to debug log in all these branches, adjust warning log so it only occurs when not a resource update conflict.